libraries/extensions/ros2-bridge/python/src/typed/serialize/mod.rs

```diff
@@ -0,0 +1,172 @@
+use std::{borrow::Cow, collections::HashMap, fmt::Display};
+
+use arrow::array::{Array, ArrayRef, AsArray};
+use dora_ros2_bridge_msg_gen::types::{
+    primitives::{GenericString, NestableType},
+    MemberType,
+};
+use serde::ser::SerializeTupleStruct;
+
+use super::{TypeInfo, DUMMY_STRUCT_NAME};
+
+mod array;
+mod defaults;
+mod primitive;
+mod sequence;
+
+#[derive(Debug, Clone)]
+pub struct TypedValue<'a> {
+    pub value: &'a ArrayRef,
+    pub type_info: &'a TypeInfo<'a>,
+}
+
+impl serde::Serialize for TypedValue<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let empty = HashMap::new();
+        let package_messages = self
+            .type_info
+            .messages
+            .get(self.type_info.package_name.as_ref())
+            .unwrap_or(&empty);
+        let message = package_messages
+            .get(self.type_info.message_name.as_ref())
+            .ok_or_else(|| {
+                error(format!(
+                    "could not find message type {}::{}",
+                    self.type_info.package_name, self.type_info.message_name
+                ))
+            })?;
```

Comment:
The operation should be pretty cheap, as it's just two HashMap lookups. But we can cache it of course if you prefer.

Reply:
Yeah, I was thinking about caching the nested types as well, but it does not seems to be doable with current definitions from `msg-gen` it seems. 

It still does feel like a good amount of lookup as we're doing it for all nested type as well at each serialization/deserialization.

Reply:
It would be doable, but I doubt that it would improve performance. The hash table lookups should be _very_ fast already, so this is something that we should not try to optimize without benchmarking/profiling.